numpy/distutils/ccompiler_opt.py

```diff
@@ -0,0 +1,1648 @@
+            if compiler_type in ("intelem", "intelemw"):
+                return "x86_64"
+            if compiler_type in ("msvc", "intel", "intelw", "intele", "mingw32"):
+                if platform.architecture()[0] == "32bit":
+                    return "x86"
+                else:
+                    return "x86_64"
+
+            return platform.machine()
+
+        compiler_info = compiler.compiler[0]
+
+        import re
+        if re.match("^(x86_64|amd64|AMD64)", compiler_info):
+            return "x86_64"
+        if re.match("^(x86|i386|i686)", compiler_info):
+            return "x86"
+        if re.match("^(powerpc|ppc)64le", compiler_info):
+            return "ppc64le"
+        if re.match("^(powerpc|ppc)64", compiler_info):
+            return "ppc64"
+        if re.match("^(arm|ARM)", compiler_info):
+            return "arm"
+        if re.match("^(aarch64|AARCH64)", compiler_info):
+            return "aarch64"
+
+        return platform.machine()
+
+    def _detect_compiler_family(self, compiler):
+        compiler_type = compiler.compiler_type
+        if compiler_type in ("unix", "mingw32", "mingw64"):
+            return "unix"
+        if compiler_type in ("intel", "intele", "intelem"):
+            return "intel"
+        if compiler_type in ("intelw", "intelemw"):
+            return "intelw"
+        return compiler_type
+
+class ccompiler_opt_features(ccompiler_opt_infa):
+
+    features_test_path = None
+
+    def __init__(self, compiler):
+        """
+        Parameters
+        ----------
+        compiler : CCompiler instance
+        """
+        ccompiler_opt_infa.__init__(self, compiler)
+
+        # 'default_baseline' : default baseline CPU features, if they are supported by C compiler
+        self._default_baseline = dict(
+            x86     = ["SSE2"],
+            x86_64  = ["SSE3"],
+            ppc64   = ["VSX"],
+            ppc64le = ["VSX2"],
+            arm     = [], # play it safe
+            aarch64 = ["ASIMD"]
+        )
+
+        # '_default_dispatch' : default dispatched CPU features, if they are supported by C compiler
+        self._default_dispatch = dict(
+            x86 = [
+                "SSE3",  "SSSE3", "SSE41", "POPCNT", "SSE42", "AVX",
+                "F16C", "XOP", "FMA4", "FMA3", "AVX2", "AVX512F", "AVX512CD",
+                "AVX512_KNL", "AVX512_KNM", "AVX512_SKX", "AVX512_CLX", "AVX512_CNL", "AVX512_ICL"
+            ],
+            x86_64  = [
+                "SSSE3", "SSE41", "POPCNT", "SSE42", "AVX",
+                "F16C", "XOP", "FMA4", "FMA3", "AVX2", "AVX512F", "AVX512CD",
+                "AVX512_KNL", "AVX512_KNM", "AVX512_SKX", "AVX512_CLX", "AVX512_CNL", "AVX512_ICL"
+            ],
+            ppc64   = ["VSX2", "VSX3"],
+            ppc64le = ["VSX3"],
+            arm     = ["NEON", "NEON_FP16", "NEON_VFPV4", "ASIMD", "ASIMDHP", "ASIMDDP", "ASIMDFHM"],
+            aarch64 = ["ASIMDHP", "ASIMDDP", "ASIMDFHM"],
+        )
```

Comment:
These could be class-level attributes (not self). Then the help for `--cpu-baseline` and `--cpu-dispatch` could use the appriate values for default, rather than None, which would go a long way to explaining what the values are for.

Reply:
This part of the code could deceive the reader so I had re-implement it in a different way.
The default value of dispatch-able features should be all the rest of the features that are not specified in the baseline. To make it more clear and easy to manage, I added special options can be used next to the CPU feature names in both parameters, explained as follows:

`MIN`: the safest features or the minimum support for CPU features supported by a wide range of users platforms.
`MAX`: all CPU features and groups the current platform and compiler support.
`NATIVE`:  still as the same but supports append and remove.
`NONE`: as `Type` or in a string, doesn't enable any features
`+/-` : append and remove operators to a certain feature or option. no need actually for '+'  since its the default.
examples: 
```BASH
--cpu-baseline = "min avx2 +AVX512F" 
--cpu-dispatch = "max" # enable all available features
--cpu-baseline = "native" # enable host or native features
--cpu-dispatch = "max -xop -fma4" # enable all available features and disable `AMD` legacy features"
--cpu-baseline  = "none" # disable all features
--cpu-dispatch = "none" # disable all features
--cpu-baseline = "AVX" 
--cpu-dispatch = "AVX2 AVX512F"
````
Now the class can easily check the cpu-baseline features first and then skip these features from the cpu-dispatch when it found, this behavior can be detected from the logs.

Also, I changed the default values for the baseline to  "min" and dispatch became "max -xop -fma4".
The values of `min` are now class-level attribute, inside a new class called `Config`.
For `None` just like I explained above became an option to disable features instead of enabling the default